### PR TITLE
(add_new command notification manager) Add new feature, notification …

### PR DESCRIPTION
**Pull request by JKtrigger**
Manager send email to customers.

### CODE REVIEW

File: dcf/management/commands/notification_manager.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+from proj.settings import DCF_ITEM_MAX_AGE_DAYS, EMAIL_HOST_USER
+from dcf.models import Item
+
+
+INT_DAYS_BEFORE_START_NOTIFICATION = 2  # before 2 day
+
+NOTIFICATION_DAY = dt.today() - timedelta(
+    days=(INT_DAYS_BEFORE_START_NOTIFICATION- (DCF_ITEM_MAX_AGE_DAYS or 0)))
+NON_ACTIVE_DAY = dt.now() - timedelta(days=-DCF_ITEM_MAX_AGE_DAYS or 0)
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Should be timezone aware (so please use Django internal now() function).

File: proj/settings.py [hunk 2]
```diff
@@ -9,6 +9,19 @@
     ('Admin', 'admin@example.com'),
 )
 
+DCF_ITEM_MAX_AGE_DAYS = 90
+
+
+EMAIL_HOST = 'smtp.serverhost.com'
```
Comment (inoks) [inline proj/settings.py]: No need to specify EMAIL settings, as it's user responsibility to provide proper EMAIL BACKEND.

File: dcf/models.py [hunk 3]
```diff
@@ -12,7 +12,8 @@
 
 class CustomUser(AbstractUser):
     phone = models.CharField(_('phone'), max_length=30, null=True, blank=True)
-    receive_news = models.BooleanField(_('receive news'), default=True, db_index=True)
+    receive_news = models.BooleanField(
+        _('receive news'), default=True, db_index=True)
```
Comment (inoks) [inline dcf/models.py]: What was changed here?

File: dcf/management/commands/notification_manager.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+# -*- coding:utf-8 -*-
+
+from datetime import datetime as dt, timedelta
+from socket import error as socket_error
+
+from django.core.management.base import BaseCommand
+
+from proj.settings import DCF_ITEM_MAX_AGE_DAYS, EMAIL_HOST_USER
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Should use Django native way for accessing settings values

File: dcf/management/commands/notification_manager.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        """ Sending messages """
+        if DCF_ITEM_MAX_AGE_DAYS is None:
+            return
+        items = Item.objects.select_related('user')
+        for item in items.filter(
+                posted__gte=NOTIFICATION_DAY.date(),
+                posted__lt=NON_ACTIVE_DAY.date(),
+        ).iterator():
+            try:
+                item.user.email_user(
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Can find `email_user` definition, should use `send_mail()` instead.

File: dcf/management/commands/notification_manager.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    def handle(self, *args, **options):
+        """ Sending messages """
+        if DCF_ITEM_MAX_AGE_DAYS is None:
+            return
+        items = Item.objects.select_related('user')
+        for item in items.filter(
+                posted__gte=NOTIFICATION_DAY.date(),
+                posted__lt=NON_ACTIVE_DAY.date(),
+        ).iterator():
+            try:
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: No need to use `try/except` here, use `fail_silently =True` instead.

File: dcf/management/commands/notification_manager.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    def handle(self, *args, **options):
+        """ Sending messages """
+        if DCF_ITEM_MAX_AGE_DAYS is None:
+            return
+        items = Item.objects.select_related('user')
+        for item in items.filter(
+                posted__gte=NOTIFICATION_DAY.date(),
+                posted__lt=NON_ACTIVE_DAY.date(),
+        ).iterator():
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Why do you use `iterator()` here?

File: dcf/management/commands/notification_manager.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            try:
+                item.user.email_user(
+                    # subject
+                    '{slug}:{title} will deactivated '
+                    'after 2 day'.format(**item.__dict__),
+                    # message
+                    'Your article "{slug}:{title} '
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Email message/body should use translation and better to use separate template (at least for body).

File: dcf/management/commands/notification_manager.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+                    # from_email
+                    EMAIL_HOST_USER,
+                )
+            except socket_error:
+                # TODO : Ask about the behavior in this case
+                pass
+        # Update message status
+        items.filter(
+            posted__gte=NON_ACTIVE_DAY.date()
+        ).filter(is_active=True).update(is_active=False)
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Should also notify user if item was deactivated.

File: dcf/management/commands/notification_manager.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    Command for status change and notification users
+    """
+    help = 'Check status per day '
+
+    def handle(self, *args, **options):
+        """ Sending messages """
+        if DCF_ITEM_MAX_AGE_DAYS is None:
+            return
+        items = Item.objects.select_related('user')
+        for item in items.filter(
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: User should not receive more than one email if he has multiple items with fulfilled conditions.

File: dcf/management/commands/notification_manager.py (hunk 4 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/management/commands/notification_manager.py (hunk 1 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/management/commands/notification_manager.py (hunk 10 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/management/commands/notification_manager.py (hunk 7 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/management/commands/notification_manager.py (hunk 6 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/management/commands/notification_manager.py (hunk 5 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/management/commands/notification_manager.py (hunk 8 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/management/commands/notification_manager.py (hunk 9 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/models.py (hunk 3 above)
Comment (JKtrigger) [inline dcf/models.py]: Corrected

File: proj/settings.py (hunk 2 above)
Comment (JKtrigger) [inline proj/settings.py]: Corrected

File: dcf/tests/test_mail.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+# -*- coding:utf-8 -*-
+
+from django.core import mail
+from django.test import TestCase
+
+
+class EmailSendTest(TestCase):
```
Comment (inoks) [inline dcf/tests/test_mail.py]: What is the purpose of testing email functionality here?

File: dcf/management/commands/notification_manager.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+                    'Your article {slug}:{title} has been '
+                    'moved to the archive'.format(**item.__dict__)
+                ),
+                # from_email
+                settings.EMAIL_HOST_USER,
+                # to user email
+                turn_off_item.user.email,
+                fail_silently=True
+            )
+            queryset_turn_off_items.update(is_active=False)
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Seems wrong place/indent?

File: dcf/management/commands/notification_manager.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    """
+    help = 'Check status per day '
+
+    def handle(self, *args, **options):
+        """ Sending messages """
+        if settings.DCF_ITEM_MAX_AGE_DAYS is None:
+            return
+        items = Item.objects.select_related('user')
+        for item in items.filter(
+                posted__range=notification_period,
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: If user have more that one outdated item you will send multiple emails, should be only one with all outdated items for that user.

File: dcf/management/commands/notification_manager.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+from django.utils.translation import ugettext as _
+from django.conf import settings
+from django.core.management.base import BaseCommand
+from django.core.mail import send_mail
+
+
+from dcf.models import Item
+
+DAY = 1
+INT_DAYS_BEFORE_START_NOTIFICATION = 2 * DAY
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Should be moved to settings.

File: dcf/management/commands/notification_manager.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+timedelta_end_notification_period = timedelta_start_notification_period + (
+    timedelta(days=DAY))
+
+notification_period = [
+    dt.now() + timedelta_start_notification_period,
+    dt.now() + timedelta_end_notification_period]
+
+NON_ACTIVE_DAY = dt.now() - timedelta(
+    days=-settings.DCF_ITEM_MAX_AGE_DAYS or 0)
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: counter-intuitive, should be simplified.

File: dcf/management/commands/notification_manager.py (hunk 12 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]:
```        for turn_off_item in queryset_turn_off_items:
            send_mail(
                # subject
                _(
                    '{slug}:{title} is already '
                    'deactivated'.format(**item.__dict__)
                ),
                # message
                _(
                    'Your article {slug}:{title} has been '
                    'moved to the archive'.format(**item.__dict__)
                ),
                # from_email
                settings.EMAIL_HOST_USER,
                # to user email
                turn_off_item.user.email,
                fail_silently=True
            )
            queryset_turn_off_items.update(is_active=False)```

Comment (inoks) [inline dcf/management/commands/notification_manager.py]: You don't need to run this query each time you send email message, aren't you?

File: dcf/tests/test_mail.py (hunk 11 above)
Comment (JKtrigger) [inline dcf/tests/test_mail.py]: Corrected

File: dcf/management/commands/notification_manager.py (hunk 14 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/management/commands/notification_manager.py (hunk 15 above)
Comment (JKtrigger) [inline dcf/management/commands/notification_manager.py]: Corrected

File: dcf/tests/test_notification_manger.py [hunk 16]
```diff
@@ -0,0 +1,14 @@
+# -*- coding:utf-8 -*-
```
Comment (inoks) [inline dcf/tests/test_notification_manger.py]: You should test new only functionality. No less, no more.
I expect your `TestCase` to cover:
- Message being sent for item which will be closed soon
- Item being deactivated if it was posted/updated more than `DCF_ITEM_MAX_AGE_DAYS` days before and message was sent to user
- Only one message was sent in each case

File: proj/settings.py [hunk 17]
```diff
@@ -9,6 +9,10 @@
     ('Admin', 'admin@example.com'),
 )
 
+DAY = 1
```
Comment (inoks) [inline proj/settings.py]: Why?

File: dcf/management/commands/notification_manager.py [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+    Command for status change and notification users
+    """
+    help = 'Check status per day '
+
+    def handle(self, *args, **options):
+        """ Sending messages """
+        if settings.DCF_ITEM_MAX_AGE_DAYS is None:
+            return
+        custom_users_with_out_of_date_items = CustomUser.objects.filter(
+            id__in=Item.objects.filter(
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Can be simplified.

File: dcf/management/commands/notification_manager.py [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+                    # msg_template,
+                    'Your articles: {} has been '
+                    'moved to the archive.',
+                    # from_email
+                    admin_mail,
+                    # to user email
+                    custom_user.email,
+                    fail_silently=True
+                )
+                list_of_dicts_where_posted_older_non_active_day.update(
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Should be executed once for all queryset, in order to limit DB hits.

File: dcf/management/commands/notification_manager.py [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+        if settings.DCF_ITEM_MAX_AGE_DAYS is None:
+            return
+        custom_users_with_out_of_date_items = CustomUser.objects.filter(
+            id__in=Item.objects.filter(
+                Q(posted__range=notification_period) |
+                Q(posted__lte=non_active_day.date()),
+                is_active=True
+            ).values('user_id'))
+
+        for custom_user in custom_users_with_out_of_date_items:
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Data can be gathered in one query.

File: dcf/management/commands/notification_manager.py [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+
+timedelta_deadline_transfer_to_archive = timedelta(
+    days=-DCF_ITEM_MAX_AGE_DAYS)
+
+non_active_day = dt.now() + timedelta_deadline_transfer_to_archive
+
+admin_mail = settings.ADMINS[0][1]
+
+
+def summary_body_send_message(
```
Comment (inoks) [inline dcf/management/commands/notification_manager.py]: Email message body should be generated inside Django Template.

**Pull request closed without merging**